consensus/reactor.go

```diff
@@ -964,7 +963,8 @@ func (ps *PeerState) SetHasProposal(proposal *types.Proposal) {
 	if ps.PRS.Height != proposal.Height || ps.PRS.Round != proposal.Round {
 		return
 	}
-	if ps.PRS.Proposal {
+	// ps.PRS.ProposalBlockParts is set due to NewValidBlockMessage
+	if ps.PRS.Proposal || ps.PRS.ProposalBlockParts != nil {
```

Comment:
The key part here is `ps.PRS.ProposalBlockParts != nil`. In this case we should't update PRS.ProposalBlockPartsHeader as it is already set (either by previous Proposal message or NewValidBlockMessage). But we might still want to set PRS.Proposal to true in case it is not already set. So maybe we should split this into 2 conditions. Regarding your last point, negation is  !PRS.Proposal && PRS.ProposalBlockParts == nil and this corresponds to the case that we are receiving Proposal for the first time and we haven't received NewValidBlockMessage yet.